x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java

```diff
@@ -574,6 +572,117 @@ public void testConvertingSearchableSnapshotFromFullToPartial() throws Exception
+        createPolicy(
+            client(),
+            policyFrozen,
+            null,
+            null,
+            new Phase(
+                "cold",
+                TimeValue.ZERO,
+                singletonMap(SearchableSnapshotAction.NAME, new SearchableSnapshotAction(snapshotRepo, randomBoolean()))
+            ),
+            new Phase(
+                "frozen",
+                TimeValue.ZERO,
+                singletonMap(SearchableSnapshotAction.NAME, new SearchableSnapshotAction(snapshotRepo, randomBoolean()))
+            ),
+            null
+        );
+
+        createIndex(index, Settings.EMPTY);
+        ensureGreen(index);
+        indexDocument(client(), index, true);
+
+        // enable ILM after we indexed a document as otherwise ILM might sometimes run so fast the indexDocument call will fail with
+        // `index_not_found_exception`
+        updateIndexSettings(index, Settings.builder().put(LifecycleSettings.LIFECYCLE_NAME, policyCold));
+
+        final String fullMountedIndexName = SearchableSnapshotAction.FULL_RESTORED_INDEX_PREFIX + index;
+
+        assertBusy(() -> {
+            logger.info("--> waiting for [{}] to exist...", fullMountedIndexName);
+            assertTrue(indexExists(fullMountedIndexName));
+        }, 30, TimeUnit.SECONDS);
+
+        assertBusy(() -> {
+            Step.StepKey stepKeyForIndex = getStepKeyForIndex(client(), fullMountedIndexName);
+            assertThat(stepKeyForIndex.phase(), is("cold"));
+            assertThat(stepKeyForIndex.name(), is(PhaseCompleteStep.NAME));
+        }, 30, TimeUnit.SECONDS);
+
+        // remove ILM
+        {
+            Request request = new Request("POST", "/" + fullMountedIndexName + "/_ilm/remove");
+            Map<String, Object> responseMap = responseAsMap(client().performRequest(request));
+            assertThat(responseMap.get("has_failures"), is(false));
+        }
+        // add cold-frozen
+        updateIndexSettings(index, Settings.builder().put(LifecycleSettings.LIFECYCLE_NAME, policyFrozen));
+        String partiallyMountedIndexName = SearchableSnapshotAction.PARTIAL_RESTORED_INDEX_PREFIX + fullMountedIndexName;
+        assertBusy(() -> {
+            logger.info("--> waiting for [{}] to exist...", partiallyMountedIndexName);
+            assertTrue(indexExists(partiallyMountedIndexName));
+        }, 30, TimeUnit.SECONDS);
+
+        assertBusy(() -> {
+            Step.StepKey stepKeyForIndex = getStepKeyForIndex(client(), partiallyMountedIndexName);
+            assertThat(stepKeyForIndex.phase(), is("frozen"));
+            assertThat(stepKeyForIndex.name(), is(PhaseCompleteStep.NAME));
+        }, 30, TimeUnit.SECONDS);
+
+        // Ensure the searchable snapshot is not deleted when the index was deleted because it was not created by this
+        // policy. We add the delete phase now to ensure that the index will not be deleted before we verify the above
+        // assertions
```

Comment:
Sorry for the random comment here - but I'm a bit unsure of this invariant. I'm wondering if it creates unexpected behaviour. The delete phase already has a flag that says "delete snapshot" which has the purpose of allowing our users to delete or skip the deletion of the searchable snapshot that backs a mounted index. 

This new behaviour ignores that flag (in some cases). I don't think we should, I think we should delete the snapshot if the `delete_searchable_snapshot` flag is `true`) https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-delete.html#ilm-delete-options

Reply:
Hm, that's not how interpreted the docs:

> Deletes the searchable snapshot **created** in a previous phase

Since we did not create the snapshot while executing this policy, we are not going to delete it. The same happens now when we skip the searchable snapshot action and the snapshot was not created by this policy (aka the lifecycle execution does not have a record of this snapshot).

I chose not to delete it for the following reasons:
- It feels to me more inline with the documentation, assuming the `created` word has the significance I thought it had.
- This is how skip searchable snapshot works now when the snapshot was not created during this policy execution.
- And finally, in the scenario that a user has made a backup of their index and then decided to mount it and add an ILM policy to it, we won't delete the backup itself but just the searchable snapshot index. I think some users might see it as an unwanted side-effect.

I understand we might have unwanted searchable snapshots in this way, but the user can always delete them later, but they cannot retrieve the lost data if we delete them by accident.

Does this make sense to you?

Reply:
++ thanks for the details. Let's keep it as is and update the docs to reflect that some snapshots could remain dangling.